e2e/playwright/conftest.py

```diff
@@ -0,0 +1,365 @@
+        self._stdout_file = TemporaryFile("w+")
+        print(f"Running: {self.args}")
+        self._proc = subprocess.Popen(
+            self.args,
+            cwd=self.cwd,
+            stdout=self._stdout_file,
+            stderr=subprocess.STDOUT,
+            text=True,
+            env={**os.environ.copy(), **self.env} if self.env else None,
+        )
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self._proc is not None:
+            self._proc.terminate()
+            self._proc = None
+        if self._stdout_file is not None:
+            self._stdout_file.close()
+            self._stdout_file = None
+
+
+def is_port_available(port: int, host: str) -> bool:
+    """Check if a port is available on the given host."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        return sock.connect_ex((host, port)) != 0
+
+
+def find_available_port(
+    min_port: int = 10000,
+    max_port: int = 65535,
+    max_tries: int = 50,
+    host: str = "localhost",
+) -> int:
+    """Find an available port on the given host."""
+    for _ in range(max_tries):
+        selected_port = randint(min_port, max_port)
+        if is_port_available(selected_port, host):
+            return selected_port
+    raise RuntimeError("Unable to find an available port.")
+
+
+def is_app_server_running(port: int, host: str = "localhost") -> bool:
+    """Check if the app server is running."""
+    try:
+        return requests.get(f"http://{host}:{port}/", timeout=1).status_code == 200
+    except Exception:
+        return False
+
+
+def wait_for_app_server_to_start(port: int, timeout: int = 5) -> bool:
+    """Wait for the app server to start."""
+    print(f"Waiting for app to start... {port}")
+    start_time = time.time()
+    while not is_app_server_running(port):
+        time.sleep(3)
+        if time.time() - start_time > 60 * timeout:
+            return False
+    return True
+
+
+def wait_for_app_loaded(page: Page):
+    """Wait for the app to fully load."""
+    # Wait until we know the script has started.
+    page.wait_for_selector(
+        "[data-testid='stAppViewContainer']", timeout=20000, state="attached"
+    )
+
+    page.wait_for_selector(
+        "[data-testid='block-container']", timeout=20000, state="attached"
+    )
+
+    # Give the app a little more time to render everything
+    time.sleep(0.25)
+    # Check that "Please wait..." does not exist within the 'stAppViewContainer'
+    app_view_container = page.query_selector("[data-testid='stAppViewContainer']")
+    if (
+        app_view_container
+        and app_view_container.inner_text().find("Please wait...") != -1
+    ):
+        pytest.fail("Found 'Please wait...' inside the stAppViewContainer")
+    # Wait until the script is no longer running.
+    page.wait_for_selector(
+        "[data-testid='stStatusWidget']", timeout=20000, state="detached"
+    )
+    # Give the app a little more time to render everything
+    time.sleep(0.5)
```

Comment:
We have to think about what assertions to have here, because the stability of our solution depends on it. I'm not sure if we check to see if "Please wait" appears. I think it's enough that we confirm the final state, i.e. the "Running..." message will disappear.

Reply:
Yep, indeed. This part is mostly a rewrite of what we currently do in Cypress. But maybe there is a better way to detect that a Streamlit app has rendered everything. 

> the "Running..." message will disappear.

This is probably referring to the `Running...` text next to the running man (stStatusWidget) icon?

Reply:
Unfortunately, without the sleep it sometimes hasn't finished rendering certain elements. I have seen this happening with the dataframe canvas and assume that it might also happen with charts or other more complex  objects that have a slight delay in it's rendering